### PR TITLE
JS: Use async generator instead of callbacks + add tests

### PR DESCRIPTION
**Pull request by yuvipanda**
JS now also supports writing async code with [Asnc Generators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AsyncGenerator) which work very much like how async code would in Python. This allows us to write code that flows better, is more familiar to python folks, and avoids messy callback hell.

This code is a *straight* port of the existing callback based code to use an Async Generator instead. The EventIterator library is a very helpful convenience function here, allowing us to translate the callbacks from EventSource into an async generator without having to write a queue of promises ourselves. The iterator is stopped exactly the same way the callbacks were earlier stopped (whenever .stop() is called).

All the callbacks have been switched over to a switch / case statement instead. We can get cleaner and use exceptions to handle failures, instead of the 'failure' case, but that can come in a later PR so this is a purely mechanical refactoring from one async method to another.

An appropriate test for this is also added. To fully test the EventSource flow, we create a temporary web server that will serve a real captured eventsource response, with 1ms gaps between lines to fully simulate how this would work in reality. This gives us much more confidence about how this would work than mocking.

With this, @jupyterhub/binder-client now has 100% unit test coverage!

Ref https://github.com/jupyterhub/binderhub/issues/774

Here is a video of me testing this. I tested both something that worked, and an image that is known to fail (RStudio on ARM).

https://github.com/jupyterhub/binderhub/assets/30430/2ce22011-ee53-4c27-b9b8-1ee27ba13c63

### CODE REVIEW

File: binderhub/static/js/index.js [hunk 1]
```diff
@@ -52,67 +52,67 @@ function build(providerSpec, log, fitAddon, path, pathType) {
...
-    // If we fail for any reason, show an error message and logs
-    updateFavicon(BASE_URL + "favicon_fail.ico");
-    log.show();
-    if ($("div#loader-text").length > 0) {
-      $("#loader").addClass("error");
-      $("div#loader-text p.launching").html(
-        "Error loading " + spec + "!<br /> See logs below for details.",
-      );
+    switch (data.phase) {
+      case "waiting":
```
Comment (consideRatio) [inline binderhub/static/js/index.js:65]: Apparently the content under case should be put in a curly bracket

https://docs.w3cub.com/eslint/rules/no-case-declarations

Comment (yuvipanda) [reply binderhub/static/js/index.js:65]: oooh, fun lexical scope thingies. I'll add.

Comment (yuvipanda) [reply binderhub/static/js/index.js]: Fixed

**Comment by yuvipanda** [comment]
I added a screen recording of me testing this as well.

File: binderhub/static/js/index.js [hunk 2]
```diff
@@ -170,21 +177,21 @@ function indexMain() {
     return false;
   });
 
-  $("#build-form").submit(function () {
+  $("#build-form").submit(async function (e) {
+    e.preventDefault();
```
Comment (yuvipanda) [inline binderhub/static/js/index.js:181]: This was necessary as `return false` doesn't work from async functions. Same effect

Comment (consideRatio) [reply binderhub/static/js/index.js:181]: I figure no - that the event to submit this form isnt considered by parent elements, but I'm not confident reviewing this in general.

Comment (yuvipanda) [reply binderhub/static/js/index.js:181]: I don't think it's relevant.

**Review by consideRatio — APPROVED** [review]

**Comment by consideRatio** [comment]
I'm not soo comfortable reviewing the JS code, but I think this makes sense and I didn't spot any issues are carefully going through this PR.

Thank you @yuvipanda for working this! This seems great!!

**Pull request merged**